### PR TITLE
When underwater, attenuate scene colour to model outscattering of light

### PR DESCRIPTION
**Pull request by huwb**
## Status

Ready to merge

## What & why

Right now when camera is deep underwater and close to the seabed, the seabed still appears bright and it looks weird. I believe this is the primary issue reported in #318 . This change multiplies the scene colour by a term based on camera depth to model light getting scattered out and diminishing with depth.

Without this change - cam deep underwater near terrain:

image

With this change - light diminishes naturally and will go to black if deep enough:

image

## How

Multiplies scene colour by term based on camera depth to model light getting scattered out and diminishing with depth.

## Testing

Ran in win64 editor. Ran win64 standalone build.

## Issues and limitations

None that i'm aware of. I suspected a few potential issues but it seems to work very well.

### CODE REVIEW

**Comment by daleeidd** [comment]
Although it does work well for the sun, it also affects other lights. It wouldn't be possible to do a deep underwater scene with a torch for example.

**Comment by huwb** [comment]
Good point! Hadn't thought about that.

I'm not sure if there's a solution to that, as all the lighting is baked in at this point. 

I guess an alternative would be to use a script that actually scales down the directional light intensity. Might be just crazy enough to work. I'll post back when I've tried it.

**Comment by moosichu** [comment]
Would the idea be to have a script which reduces the intensity of directional lights based on their distance from the camera based on the depth-fog-density?

**Comment by huwb** [comment]
That would be an interesting test. Maybe just the sunlight (?) as assigned to the oceanrenderer. If you want to try it out feel free, otherwise i will later.

**Comment by huwb** [comment]
Just reread your question, no it would just be reducing the directional light intensity based on camera depth compare to sea level. So moving the changes in this PR from the hlsl into a c# script:

```
				// Darken the background to model light outscattered as it passes from sun/env through water to the background surface
				half camDepth = max(_OceanCenterPosWorld.y - _WorldSpaceCameraPos.y, 0.0);
				sceneColour *= exp(-_DepthFogDensity.xyz * camDepth * DEPTH_OUTSCATTER_CONSTANT);
```

**Comment by huwb** [comment]
Well think i found a pretty horrible unity bug. the tangent vector was using a transform node with the space set to Absolute World. i didnt think this mattered as its set to Direction so position should not influence it.

image

I guess what it is doing is a translation to convert from absolute world to world, and then treating the result as a direction, which meant it would shift around depending on camera location.

Setting the space to World seemed to be enough for me, but ill try it again at SRP level now.

**Comment by huwb** [comment]
Not sure if you decided to pick this up @moosichu , apologies if so, but i had a bit more time and had a stab at this.

I multiply down the primary light intensity, and also the ambient light intensity. Both are required to ramp down the light.

I'm not sure how i feel about changing the light in the scene, but it seems to work and i cant put my finger on why its not a good idea, so maybe it will work.

The one thing i dont love is having to read the fog values off the material, which means if the value on the material is changed it wont be picked up. I also don't love that it runs per frame but i guess thats ok.

The last thing is that if underwater is not enabled then we wouldnt want this to run, potentially adding yet another user configurable option. This can probably be eliminated by moving the code to UnderwaterEffect.cs. I'll look at this if we decide to proceed with this approach.

I'll test in the 'RPs next.

**Comment by daleeidd** [comment]
Having it as an option is a good idea since most weather assets override the sun values and they will have their own way of managing it. Is it possible to expose the multiplier as a property? Then we could have our own component in crest examples which applies the values to sun and ambient. This might work better for downstream too due to lighting differences.

**Comment by huwb** [comment]
Ah good idea, I'll try

**Comment by huwb** [comment]
Ok the mechanism i was using to scale down the ambient lighting - *RenderSettings.ambientIntensity* - does nothing at all in URP. I looked through the lighting code and didnt spot any reference to a scalar on the lighting.

In this case i'm inclined to stick with the original solution - just scale the background colour - and call not supporting local lights a limitation. Its a shame but i dont see a way around it. Let me know what you think.

**Comment by daleeidd** [comment]
It works for me but only when a _Skybox Material_ is set and _Environmental Lighting > Source_ is set to _Skybox_. You should see the _Intensity Multiplier_ slider when the settings are correct.

Also, we will have to disable the fog when underwater. Either by disabling it, or countering it in the underwater curtain. HDRP (and SRP once it supports it) can simply use the volume framework to blend out the fog once deep enough underwater. Fog would affect the underwater effects regardless of our solution for outer scattering.

The last issue is to do with baked lighting. If the lighting is baked when the skybox exposure is set to one, the ambient probe will have some colour even when the intensity multiplier is set to 0. There is nothing we can do about that it seems until Unity supports realtime GI in SRP.

Also, if we expose the depth fog colour, as well as the multiplier, then users will be able to add scripts to their lights to get the right colour.

I'm happy to work on this if you want to work on other things?

**Comment by huwb** [comment]
Thanks for the dilligence

Interesting that we're using *Gradient* as the source in that scene, im doubtful whether this should be the case. i'm guessing we should probably just switch it to skybox.

> For HDRP we would have to blend out (or blend in 0 exposure sky cos of default settings or disable those default settings) the atmosphere volume in addition to the directional light.

Ok makes sense

I think this would be really cool. Lets chat on discord about short-term priorities.

**Comment by daleeidd** [comment]
I made it a seperate component. I have confirmed that it works with URP. 

I added a spotlight to the end of the camera. Let me know if you want it kept or not.

As for HDRP, I have worked out a solution which works. It just requires an additional global volume which we will blend in which overrides indirect lighting (and anything else we need) as zero.

**Comment by daleeidd** [comment]
I was thinking one thing we could add is a light intensity parameter to the ocean material which we could increase with depth to counter the unnatural change in lighting. It didn't make much of a difference though.

**Comment by huwb** [comment]
Tested on birp, looks really cool.

I'm not sure we should leave a spotlight on the camera by default, for perf and misc. I'll think about it. We can probably demo it in other ways, like little glowing red mines?

**Comment by huwb** [comment]
I didnt understand the parameter note. What is the unnatural change? In any case sounds like it didnt make an impact.

**Comment by huwb** [comment]
Would it be worth moving the changes in OceanRenderer into the underwater script? I think i'd vote for this - if we're going to add a new script might as well put all the changes there, perhaps?

There could conceivably be something else that cares about the depth multiplier in the future, but i dont know what that would be, and we could always move it back.

File: crest/Assets/Crest/Crest-Examples/Shared/Scripts/UnderwaterEnvironmentalLighting.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+﻿// Crest Ocean System
+
+// This file is subject to the MIT License as seen in the root of this folder structure (LICENSE)
+
+using UnityEngine;
+
+namespace Crest
+{
+    /// <summary>
+    /// Provides outer scattering based on the camera's underwater depth. It scales down environmental lighting
```
Comment (huwb) [inline crest/Assets/Crest/Crest-Examples/Shared/Scripts/UnderwaterEnvironmentalLighting.cs]: Minor typo - outer scattering should be 'outscattering'

**Comment by daleeidd** [comment]
> I'm not sure we should leave a spotlight on the camera by default, for perf and misc. I'll think about it. We can probably demo it in other ways, like little glowing red mines?

Sounds good. Done.

> Would it be worth moving the changes in OceanRenderer into the underwater script?

Sure thing. Done.

> I didnt understand the parameter note. What is the unnatural change? In any case sounds like it didnt make an impact.

We are reducing the lighting which affects the ocean surface and depth fog. I thought it would make the surface darken sooner than it should. But the impact was minor (I was sticking between 1 and 2 multiplier).

I also ported the out-scattering constant over.

File: crest/Assets/Crest/Crest-Examples/Shared/Scripts/UnderwaterEnvironmentalLighting.cs (hunk 1 above)
Comment (daleeidd) [reply crest/Assets/Crest/Crest-Examples/Shared/Scripts/UnderwaterEnvironmentalLighting.cs]: Thanks for the correction. For some reason it has been outer scattering in my mind this whole time lol

**Comment by huwb** [comment]
Happy for this to merge when you are. Nice one

**Comment by daleeidd** [comment]
Your additions looks great. Merged

**Pull request merged**